aten/src/ATen/core/ivalue_inl.h

```diff
@@ -430,20 +430,18 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   // Expose the default implementation so that external ones can defer to it.
   static std::vector<std::reference_wrapper<const at::DataPtr>>
   defaultDataPtrExtractor(const at::IValue& value) {
-    // FIXME Should we support more types than just tensors and tensor lists?
-    TORCH_INTERNAL_ASSERT(
-        value.isTensorList() || value.isTensor(),
-        "the future value must be either a tensor list or a tensor.");
-    at::Tensor tensor;
-    if (value.isTensorList()) {
-      const auto tensors = value.toTensorVector();
-      TORCH_INTERNAL_ASSERT(tensors.size() == 1, "expected exactly 1 tensor");
-      tensor = tensors[0];
-    } else {
-      tensor = value.toTensor();
+    at::IValue::HashAliasedIValues sub_values;
+    // Prefer getSubValues() over visit() as the latter is a silent no-op for
+    // some unsupported types, whereas the former at least fails loudly.
+    value.getSubValues(sub_values);
```

Comment:
To clarify this, it seems that getSubValues triggers an assert for these types:
https://github.com/pytorch/pytorch/blob/8b248af35d43c97d0e437f6f4ff0fbd4da5700c8/aten/src/ATen/core/ivalue.cpp#L202-L206
Whereas visit just ignores such values but succeeds silently. I must admit that I expected getSubValues and visit to behave consistently (and perhaps the former to be implemented through the latter). Is there any reason for that?

Also, are there fundamental reasons for those types to not be supported? Are they planned in the future?

Are there better ways to extract DataPtrs from an IValue? One idea I had was pickling it, but I'm worried about the performance implications.

Reply:
cc @wanchaol :)

Reply:
> Are there better ways to extract DataPtrs from an IValue? One idea I had was pickling it, but I'm worried about the performance implications.

I assume you are concerned about nested data structures with Tensors? It should be fine for `ProcessGroupNCCL` and futures returned by RPC, but might need a recursive/pickle-like solution for general `CudaFuture` objects returned by then?

Reply:
Two cases that I'm not sure are supported in this version of the code are mixed-type lists (e.g., the user having a callback that does `return [42, tensor]`) and custom user-defined classes that contain tensors as some fields.

Also, I am not sure how this applies to RPC, because in that case the IValue held by the future will be a `Message`, which will contain the result of the remote user function as a field, and that value could be of any type, including a mixed-type lists and user-defined classes. Right?

Reply:
As a general comment, if `getSubValues()` or `visit()` was able to support PyObjects (and properly acquire the GIL and such), then we wouldn't need all that DataPtr extractor craziness. That would by far be the ideal scenario. I don't know how much effort is required to get there...

Reply:
> admit that I expected getSubValues and visit to behave consistently (and perhaps the former to be implemented through the latter). Is there any reason for that?

semantic-wise it seems these two functions are a bit different, but yeah i agree they could be rewrite together. cc @bzinodev  who implemented those two functions.


> are there fundamental reasons for those types to not be supported? Are they planned in the future?

I think it's a matter of semantic undefined for those kind of types, i.e. what does it mean to get a subvalues of a `FutureType` ivalue? there might not be subvalues inside if the future is not completed. If we could defined well on those semantics, we could add those. 

> Are there better ways to extract DataPtrs from an IValue?

I think what you need here is a "flatten" function which flattens all different types of objects so that you could iterate them and get all the tensor data ptrs for those "tensor" ivalues. Your approach looks good by using `getSubValues`, though we still need to define what it should behave for those unsupported types if we want to expand it.

 > As a general comment, if getSubValues() or visit() was able to support PyObjects (and properly acquire the GIL and such), then we wouldn't need all that DataPtr extractor craziness.

Can you explain a bit more when/why we need to extract PyObjects to get DataPtrs? not sure if I understood fully on why supporting PyObjects would remove those craziness. But you want a python land to iterate the ivalues, you can possibly define those inside https://github.com/pytorch/pytorch/blob/2e9d6d99be52cdd2f27512ff357e8aae7f4d3d78/torch%2Fcsrc%2Fjit%2Fpython%2Fpython_ivalue.h#L0-L1

Reply:
Thanks for your answers!

> I think it's a matter of semantic undefined for those kind of types

Good point, makes sense! I'd say I'm totally fine with only supporting a (reasonable) subset of types in CUDAFuture. The remaining concern is how can we effectively communicate this to the user? Both in terms of how do we raise a meaningful error, and do we explain it in the docs? Do you have suggestions?

> Can you explain a bit more when/why we need to extract PyObjects to get DataPtrs?

Sorry, it's the other way around: I need to extract DataPtrs from PyObjects. The reason is that the value of a "follow-up" future could be produced by a user-provided Python function, in which case the type of the IValue would be PyObject, right?

Anyways, thanks a lot for pointing me to python_ivalue.h! I hadn't seen it before and it gave me an idea to fix this properly. I'll try to implement and send out a PR, and maybe we can discuss further over there.

Reply:
Done this in #48840.